### DISCUSSION TITLE
Wasm build fails with "Parameter types of pinvoke callback method 'System.Int32 SetMinMaxSize(System.Void*, Avalonia.Native.Interop.AvnSize, Avalonia.Native.Interop.AvnSize)' needs to be blittable."

### DISCUSSION THREAD

**Discussion by saibot75** [AvaloniaUI/Avalonia#9264] [Q&A]
When building my wasm project I get the following build error for 11.0.0-preview2:

C:\Program Files\dotnet\packs\Microsoft.NET.Runtime.WebAssembly.Sdk\6.0.10\Sdk\WasmApp.Native.targets(244,5): error : Parameter types of pinvoke callback method 'System.Int32 SetMinMaxSize(System.Void*, Avalonia.Native.Interop.AvnSize, Avalonia.Native.Interop.AvnSize)' needs to be blittable.
Done building project "Project1.Web.csproj" -- FAILED.

I don't get the error when building for 0.10.18 or building for Windows desktop. Any ideas?

**Comment by timunie** [comment]
Can you try the latest nightly build? There were many things going on recently.

**Reply by saibot75** [reply]
I cannot find any nightly builds for 11.0.0-preview2, maybe they are in a different source?
I only see 0.10.999 builds at nuget.avaloniaui.net/repository/avalonia-nightly/index.json.

**Reply by timunie** [reply]
I'm not sure about the current name of the nightly builds. But here are the instructions: https://github.com/AvaloniaUI/Avalonia/wiki/Using-nightly-build-feed

**Reply by timunie** [reply]
nightly builds seem to be broken. 🤔

**Reply by timunie** [reply]
fixed: https://github.com/AvaloniaUI/Avalonia/wiki/Using-nightly-build-feed 

please update your nuget source for nightly builds

**Reply by saibot75** [reply]
I've updated the package source and got the latest nightly 11 build. Unfortunately I now get other build error which I presume is caused by the new Avalonia.Web package being .net 7 (and I target 6.0). I haven't dared to update my VS 2022 to the preview version supporting .net 7 yet but I might try that to see if it solves my issue.

**Comment by maxkatz6** [comment]
You are adding "Avalonia.Desktop" package to the Web build with WASM. 
It's not compatible with Desktop packages.

**Reply by saibot75** [reply]
No, the only package I have on the WASM build is Avalonia.Web.Blazor. (And 2 Microsoft.AspNetCore.*)
![image](https://user-images.githubusercontent.com/18336459/197397698-b89f02a9-32a8-4559-a0cd-a57309473a07.png)

**Reply by saibot75** [reply]
But here I see something:
![image](https://user-images.githubusercontent.com/18336459/197399251-b3c0c22c-2c0d-448f-9848-fee3519f915c.png)
Appearantly FluentAvaloniaUI has a dependency to Avalonia.Desktop, could that be the problem? This seems to be a problem with the 2.0.0-preview2 version since I had no build problems with Avalonia 0.10.18 and FluentAvaloniaUI 1.4.4.

**Reply by maxkatz6** [reply]
Yes, it could. @amwx do we need this dependency there?

**Reply by amwx** [reply]
Right now, yes. `Avalonia.Desktop` is needed for `CoreWindow`, my custom Window class, and also is how I get the SkiaSharp reference used in my ColorPicker. Both are fixable (particularly the latter as I really should be using my own SkiaSharp reference), though fixing this for CoreWindow is quite involved.

But I guess my question now is what's calling that `SetMinMaxSize` method leading to this? Also what changed here that the issue doesn't occur with 0.10 branch but does with 11.0 - Avalonia.Desktop has always been referenced.

**Reply by maxkatz6** [reply]
> But I guess my question now is what's calling that SetMinMaxSize method leading to this

Nothing. It just exists in dependency graph. AOT compiler/optimized tries to lookup for all PInvokes in the library if everything is compatible with WASM.

> and also is how I get the SkiaSharp reference used in my ColorPicker

Browser is compatible with Skia, so you can just reference Avalonia.Skia. At least now while we are using SkiaSharp, there won't be problems with it. Or you can reference SkiaSharp directly.

**Reply by mimimiaomiao1** [reply]
Quoting avalonia-prism can cause this issue
